nipype/pipeline/engine/nodes.py

```diff
@@ -304,124 +312,118 @@ def run(self, updatehash=False):
-                    # remove old json files
-                    for filename in glob(op.join(outdir, '_0x*.json')):
-                        os.unlink(filename)
-            outdir = make_output_dir(outdir)
-            self._save_hashfile(hashfile_unfinished, hashed_inputs)
-            self.write_report(report_type='preexec', cwd=outdir)
-            savepkl(op.join(outdir, '_node.pklz'), self)
-            savepkl(op.join(outdir, '_inputs.pklz'),
-                    self.inputs.get_traitsfree())
-            try:
-                self._run_interface()
-            except:
-                os.remove(hashfile_unfinished)
-                raise
-            shutil.move(hashfile_unfinished, hashfile)
-            self.write_report(report_type='postexec', cwd=outdir)
-        else:
-            if not op.exists(op.join(outdir, '_inputs.pklz')):
-                logger.debug('%s: creating inputs file', self.name)
-                savepkl(op.join(outdir, '_inputs.pklz'),
-                        self.inputs.get_traitsfree())
-            if not op.exists(op.join(outdir, '_node.pklz')):
-                logger.debug('%s: creating node file', self.name)
-                savepkl(op.join(outdir, '_node.pklz'), self)
-            logger.debug("Hashfile exists. Skipping execution")
             self._run_interface(execute=False, updatehash=updatehash)
-        logger.debug('Finished running %s in dir: %s\n', self._id, outdir)
-        return self._result
+            logger.info('[Node] Cached "%s".', self.fullname)
+            os.chdir(cwd)
+            return self.result
+
+        # by rerunning we mean only nodes that did finish to run previously
+        json_pat = op.join(outdir, '_0x*.json')
+        json_unfinished_pat = op.join(outdir, '_0x*_unfinished.json')
+        is_mapnode = isinstance(self, MapNode)
+        need_rerun = (not is_mapnode and
+                      glob(json_pat) and not glob(json_unfinished_pat))
+        if need_rerun:
+            log_debug = config.get('logging', 'workflow_level') == 'DEBUG'
+            logger.debug('[Node] Rerunning "%s"', self.fullname)
+            if log_debug and not hash_exists:  # Lazy logging - only debug
+                exp_hash_paths = glob(json_pat)
+                if len(exp_hash_paths) == 1:
+                    split_out = split_filename(exp_hash_paths[0])
+                    exp_hash_file_base = split_out[1]
+                    exp_hash = exp_hash_file_base[len('_0x'):]
+                    logger.debug("Previous node hash = %s", exp_hash)
+                    try:
+                        prev_inputs = load_json(exp_hash_paths[0])
+                    except:
+                        pass
+                    else:
+                        logging.logdebug_dict_differences(prev_inputs,
+                                                          hashed_inputs)
+            if not force_run and str2bool(self.config['execution']['stop_on_first_rerun']):
```

Comment:
perhaps stop_on_first_rerun should change the logging from debug to info, so that people don't always have to switch to debug mode to see the differences.

Reply:
In fmriprep we have now more logging levels. We could have a VERBOSE level (more info than INFO, for instance these messages), and use DEBUG for deep debugging info.

WDTY?

Reply:
I modified a little how this works. With fmriprep it goes great (but we have the extra VERBOSE level). This way one can avoid the extremely verbose DEBUG, and still get the differences that make an interface be run.